src/bootman/bootman.c

```diff
@@ -446,6 +457,210 @@ bool boot_manager_set_uname(BootManager *self, const char *uname)
+        autofree(char) *initrd_source = NULL;
+        bool is_uefi = ((self->bootloader->get_capabilities(self) & BOOTLOADER_CAP_UEFI) ==
+                        BOOTLOADER_CAP_UEFI);
+        const char *efi_boot_dir =
+            is_uefi ? self->bootloader->get_kernel_destination(self) : NULL;
+        base_path = boot_manager_get_boot_dir((BootManager *)self);
+        if (!self || !self->initrd_freestanding_dir) {
+                return false;
+        }
+
+        /* not initrd to write */
+        if (!self->initrd_freestanding) {
+                return true;
+        }
+
+        /* if it's UEFI, then bootloader->get_kernel_dst() must return a value. */
+        if (is_uefi && !efi_boot_dir) {
+                return false;
+        }
+
+        initrd_target = string_printf("%s%s/%s",
+                                      base_path,
+                                      (is_uefi ? efi_boot_dir : ""),
+                                      self->initrd_freestanding);
+        initrd_source = string_printf("%s/%s",
+                                      self->initrd_freestanding_dir,
+                                      self->initrd_freestanding);
+        if (!cbm_files_match(initrd_source, initrd_target)) {
+                if (!copy_file_atomic(initrd_source, initrd_target, 00644)) {
+                        LOG_FATAL("Failed to install initrd %s: %s",
+                                  initrd_target,
+                                  strerror(errno));
+                        return false;
+                }
+        }
+        return true;
+}
+
+bool boot_manager_remove_initrd_freestanding(BootManager * self)
+{
+        autofree(char) *base_path = NULL;
+        autofree(char) *initrd_efi_path = NULL;
+        autofree(char) *initrd_target = NULL;
+        autofree(DIR) *initrd_dir = NULL;
+        struct dirent *ent = NULL;
+        bool is_uefi = ((self->bootloader->get_capabilities(self) & BOOTLOADER_CAP_UEFI) ==
+                        BOOTLOADER_CAP_UEFI);
+        const char *efi_boot_dir =
+            is_uefi ? self->bootloader->get_kernel_destination(self) : NULL;
+        if (!self || !self->initrd_freestanding_dir) {
+                return false;
+        }
+        /* if it's UEFI, then bootloader->get_kernel_dst() must return a value. */
+        if (is_uefi && !efi_boot_dir) {
+                return false;
+        }
+
+        base_path = boot_manager_get_boot_dir((BootManager *)self);
+
+        initrd_efi_path = string_printf("%s/%s",
+                                        base_path,
+                                        (is_uefi ? efi_boot_dir : ""));
+
+        initrd_dir = opendir(initrd_efi_path);
+        if (!initrd_dir) {
+                LOG_ERROR("Error opening %s: %s", initrd_efi_path, strerror(errno));
+                return false;
+        }
+
+        while ((ent = readdir(initrd_dir)) != NULL) {
+
+                if (!boot_manager_is_initrd_valid_name(ent->d_name)) {
```

Comment:
I'm a little bit scared of this being the only test to ensure that we are actually removing initrds we installed but considering clr-boot-manager asserts it owns this space I guess it is a sufficient guard.

Reply:
what could you suggest, probably add a folder for these initrds?

Reply:
That's probably a good idea but for now we can go with the method you have.